gcloud-java-storage/src/main/java/com/google/gcloud/storage/BlobRewriter.java

```diff
@@ -0,0 +1,203 @@
+    this.rewriteToken = builder.rewriteToken;
+    this.totalBytesRewritten = firstNonNull(builder.totalBytesRewritten, BigInteger.ZERO);
+    this.maxBytesRewrittenPerCall = builder.maxBytesRewrittenPerCall;
+    this.storageRpc = serviceOptions.storageRpc();
+  }
+
+  static class Builder {
+
+    private final StorageOptions serviceOptions;
+    private final BlobId source;
+    private final Map<StorageRpc.Option, ?> sourceOptions;
+    private final BlobInfo target;
+    private final Map<StorageRpc.Option, ?> targetOptions;
+    private BigInteger blobSize;
+    private Boolean isDone;
+    private String rewriteToken;
+    private BigInteger totalBytesRewritten;
+    private Long maxBytesRewrittenPerCall;
+
+    Builder(StorageOptions serviceOptions, BlobId source, Map<StorageRpc.Option, ?> sourceOptions,
+        BlobInfo target, Map<StorageRpc.Option, ?> targetOptions) {
+      this.serviceOptions = serviceOptions;
+      this.source = source;
+      this.sourceOptions = sourceOptions;
+      this.target = target;
+      this.targetOptions = targetOptions;
+    }
+
+    Builder blobSize(BigInteger blobSize) {
+      this.blobSize = blobSize;
+      return this;
+    }
+
+    Builder isDone(Boolean isDone) {
+      this.isDone = isDone;
+      return this;
+    }
+
+    Builder rewriteToken(String rewriteToken) {
+      this.rewriteToken = rewriteToken;
+      return this;
+    }
+
+    Builder totalBytesRewritten(BigInteger totalBytesRewritten) {
+      this.totalBytesRewritten = totalBytesRewritten;
+      return this;
+    }
+
+    Builder maxBytesRewrittenPerCall(Long maxBytesRewrittenPerCall) {
+      this.maxBytesRewrittenPerCall = maxBytesRewrittenPerCall;
+      return this;
+    }
+
+    BlobRewriter build() {
+      return new BlobRewriter(this);
+    }
+  }
+
+  static Builder builder(StorageOptions options, BlobId source,
+      Map<StorageRpc.Option, ?> sourceOpt,
+      BlobInfo target, Map<StorageRpc.Option, ?> targetOpt) {
+    return new Builder(options, source, sourceOpt, target, targetOpt);
+  }
+
+  /**
+   * Returns the id of the source blob.
+   */
+  public BlobId source() {
+    return source;
+  }
+
+  /**
+   * Returns the info for the target blob. When {@link #isDone} is {@code true} this method returns
+   * the updated information for the just written blob.
+   */
+  public BlobInfo target() {
+    return target;
+  }
+
+  /**
+   * Size of the blob being copied. Returns {@code null} until the first copy request returns.
+   */
+  public BigInteger blobSize() {
+    return blobSize;
+  }
+
+  /**
+   * Returns {@code true} of blob rewrite finished, {@code false} otherwise.
+   */
+  public Boolean isDone() {
+    return isDone;
+  }
+
+  /**
+   * Returns the token to be used to rewrite the next chunk of the blob.
+   */
+  public String rewriteToken() {
+    return rewriteToken;
+  }
+
+  /**
+   * Returns the number of bytes written. 
+   */
+  public BigInteger totalBytesRewritten() {
```

Comment:
Oh, I wonder if we should make it simpler and use long which will support a file of the size more than a [exbibyte](https://en.wikipedia.org/wiki/Exbibyte), especially consider that now max upload size is [5 TB](https://cloud.google.com/storage/docs/json_api/v1/objects/insert). 

Otherwise, we should create a bug and also update BlobInfo.size to use `BigInteger` instead of long.

/cc @BrandonY